app/scripts/controllers/user-storage/user-storage-controller.ts

```diff
@@ -0,0 +1,247 @@
+      throw new Error(
+        `${controllerName} - failed to enable profile syncing - ${errorMessage}`,
+      );
+    }
+  }
+
+  public async disableProfileSyncing(): Promise<void> {
+    const isAlreadyDisabled = !this.state.isProfileSyncingEnabled;
+    if (isAlreadyDisabled) {
+      return;
+    }
+
+    this.update((state) => {
+      state.isProfileSyncingEnabled = false;
+    });
+  }
+
+  /**
+   * Allows retrieval of stored data. Data stored is string formatted.
+   * Developers can extend the entry path and entry name through the `schema.ts` file.
+   *
+   * @param entryKey
+   * @returns the decrypted string contents found from user storage (or null if not found)
+   */
+  public async performGetStorage(
+    entryKey: UserStorageEntryKeys,
+  ): Promise<string | null> {
+    this.#assertProfileSyncingEnabled();
+    const { bearerToken, storageKey } =
+      await this.#getStorageKeyAndBearerToken();
+    const result = await getUserStorage({
+      entryKey,
+      bearerToken,
+      storageKey,
+    });
+
+    return result;
+  }
+
+  /**
+   * Allows storage of user data. Data stored must be string formatted.
+   * Developers can extend the entry path and entry name through the `schema.ts` file.
+   *
+   * @param entryKey
+   * @param value - The string data you want to store.
+   * @returns nothing. NOTE that an error is thrown if fails to store data.
+   */
+  public async performSetStorage(
+    entryKey: UserStorageEntryKeys,
+    value: string,
+  ): Promise<void> {
+    this.#assertProfileSyncingEnabled();
+    const { bearerToken, storageKey } =
+      await this.#getStorageKeyAndBearerToken();
+
+    await upsertUserStorage(value, {
+      entryKey,
+      bearerToken,
+      storageKey,
+    });
+  }
+
+  /**
+   * Retrieves the storage key, for internal use only!
+   *
+   * @returns the storage key
+   */
+  public async getStorageKey(): Promise<string> {
+    this.#assertProfileSyncingEnabled();
+    const storageKey = await this.#createStorageKey();
+    return storageKey;
+  }
+
+  #assertProfileSyncingEnabled(): void {
+    if (!this.state.isProfileSyncingEnabled) {
+      throw new Error(
+        `${controllerName}: Unable to call method, user is not authenticated`,
+      );
+    }
+  }
+
+  /**
+   * Utility to get the bearer token and storage key
+   */
+  async #getStorageKeyAndBearerToken(): Promise<{
+    bearerToken: string;
+    storageKey: string;
+  }> {
+    const bearerToken = await this.#auth.getBearerToken();
+    if (!bearerToken) {
+      throw new Error('UserStorageController - unable to get bearer token');
+    }
+    const storageKey = await this.#createStorageKey();
+
+    return { bearerToken, storageKey };
+  }
+
+  /**
+   * Rather than storing the storage key, we can compute the storage key when needed.
+   *
+   * @returns the storage key
+   */
+  async #createStorageKey(): Promise<string> {
+    const id = await this.#auth.getSessionIdentifier();
```

Comment:
It seems that the key used for encryption needs the session idenitfier. How is the session identifier deterministically generated across different installs?

Reply:
Great call out!

The session identifier here comes from our auth server. Under the hood this will be the "Auth V2 Profile Id", where we can sync multiple "logins" with the same profile.

[View Doc](https://docs.google.com/document/d/1RxuveLq6xFAOK5DQ7a5YFPOP_aZkUAxQQg7qLEM5lfU/edit#heading=h.nqeonx6gogmd)